LLama/LLamaWeights.cs

```diff
@@ -71,6 +71,23 @@ private LLamaWeights(SafeLlamaModelHandle weights)
             Metadata = weights.ReadMetadata();
         }
 
+        /// <summary>
+        /// Create from a "shared" handle
+        /// </summary>
+        /// <param name="handle"></param>
+        /// <returns></returns>
+        public static LLamaWeights FromSafeModelHandle(SafeLlamaModelHandle handle)
+        {
+            var model = new LLamaWeights(handle);
+
+            // Increment the model reference count while this weight exists.
+            // DangerousAddRef throws if it fails, so there is no need to check "success"
+            var success = false;
+            handle.DangerousAddRef(ref success);
```

Comment:
I don't see where this refcount is _decremented_?

Assuming I'm not missing it, the LLamaWeights class needs:
 - decrement it on `Dispose()` **if** this instance owns the safe handle (i.e. if it was created with this method).
 - finalizer to ensure that dispose gets called even if the object is not disposed.

Reply:
It's done as part of `UnloadModel` the reason for the split is, I didn't want to change existing semantics of code already using this. Although, now as I'm typing this out, as long as it gets aded to both disposes and constructors, it should be okay? I think this method is still needed though to create from a cached handle

Reply:
I think it'd be neater to change the existing semantics. Also it's a more general improvement, if any `LlamaWeights` (whether or not it's from the cache) can be "cloned" to another object with a shared ref-countered ownership.